File: hospital/AMarotoChivite-HospitalTest/AMarotoChivite-HospitalTest-Java/src/main/java/HospitalCollections/repositories/PatientMapRepository.java
package HospitalCollections.repositories;

import HospitalCollections.base.IHospitalRepository;
import HospitalCollections.models.Patient;
import HospitalCollections.models.PatientComparator;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.TreeMap;

import static HospitalCollections.models.PatientComparator.SortOrder.*;

public class PatientMapRepository implements IHospitalRepository {
    // El mapa ordena por defecto por la clave
    private final TreeMap<String, Patient> dataBase = new TreeMap<>();

    @Override
    public List<Patient> getAll() {
        // Pasamos a lista los valores del mapa
        return new ArrayList<>(dataBase.values());
    }

    @Override
    public Patient getById(String dniPacient) {
        return dataBase.get(dniPacient);
    }

    @Override
    public Patient save(Patient entity) {
        dataBase.put(entity.getDni(), entity);
        return entity;
    }

    @Override
    public Patient deleteById(String dni) {
        Patient patientDeleted = dataBase.get(dni);
        dataBase.remove(dni);
        return patientDeleted;
    }

    @Override
    public boolean isFull() {
        int maxSize = 10;
        return dataBase.size() >= maxSize;
    }

    @Override
    public int getNumPacients() {
        return dataBase.size();
    }

    @Override
    public int[] getNumByType() {
        int countNormal = 0;
        int countUrgent = 0;

        Collection<Patient> values = dataBase.values();

        for (Patient patient : values) {
            if (patient.getType().equals(Patient.TypePatient.NORMAL)) {
                countNormal++;
            } else if (patient.getType().equals(Patient.TypePatient.URGENT)) {
                countUrgent++;
            }
        }

        int[] counts = new int[2];
        counts[0] = countNormal;
        counts[1] = countUrgent;
        return counts;
    }

    @Override
    public List<Patient> getAllOrderByEnterDate() {
        List<Patient> orderedPatients = new ArrayList<>(dataBase.values());
        orderedPatients.sort(new PatientComparator(DATE_ENTER));
        return orderedPatients;
    }

    @Override
    public List<Patient> getAllOrderByName() {
        List<Patient> orderedPatients = new ArrayList<>(dataBase.values());
        orderedPatients.sort(new PatientComparator(NAME));
        return orderedPatients;
    }

    @Override
    public List<Patient> getAllOrderByType() {
        List<Patient> orderedPatients = new ArrayList<>(dataBase.values());
        orderedPatients.sort(new PatientComparator(TYPE));
        return orderedPatients;
    }
}

File: hospital/AMarotoChivite-HospitalTest/AMarotoChivite-HospitalTest-Kotlin/src/main/kotlin/HospitalCollections/repositories/PatientMapRepository.kt
package HospitalCollections.repositories

import HospitalCollections.base.IHospitalRepository
import HospitalCollections.models.Patient
import HospitalCollections.models.Patient.TypePatient
import java.util.*

class PatientMapRepository : IHospitalRepository {
    // El mapa ordena por defecto por la clave
    private val dataBase = TreeMap<String?, Patient?>()

    // Pasamos a lista los valores del mapa
    override fun getAll(): List<Patient?> {
        return dataBase.values.toList()
    }

    override fun getById(id: String?): Patient? {
        return dataBase[id]
    }

    override fun save(entity: Patient?): Patient {
        dataBase[entity!!.getDni()] = entity
        return entity
    }

    override fun deleteById(id: String?): Patient? {
        val patientDeleted = dataBase[id]
        dataBase.remove(id)
        return patientDeleted
    }

    override fun isFull(): Boolean {
        val maxSize = 10
        return dataBase.size >= maxSize
    }

    override val numPacients: Int
        get() = dataBase.size

    override val numByType: IntArray
        get() {
            var countNormal = 0
            var countUrgent = 0
            val values: Collection<Patient?> = dataBase.values
            for (patient in values) {
                if (patient!!.type == TypePatient.NORMAL) {
                    countNormal++
                } else if (patient.type == TypePatient.URGENT) {
                    countUrgent++
                }
            }
            val counts = IntArray(2)
            counts[0] = countNormal
            counts[1] = countUrgent
            return counts
        }

    override val allOrderByEnterDate: List<Patient?>
        get() {
            // Ordenar sobre la marcha sin comparator, con programación funcional
            val orderedPatients: List<Patient?> = ArrayList(dataBase.values)
            orderedPatients.sortedBy { it!!.getDateEnter() }
            return orderedPatients
        }

    override val allOrderByName: List<Patient?>
        get() {
            // Ordenar sobre la marcha sin comparator, con programación funcional
            val orderedPatients: List<Patient?> = ArrayList(dataBase.values)
            orderedPatients.sortedBy { it!!.getName() }
            return orderedPatients
        }

    override val allOrderByType: List<Patient?>
        get() {
            // Ordenar sobre la marcha sin comparator, con programación funcional
            val orderedPatients: List<Patient?> = ArrayList(dataBase.values)
            orderedPatients.sortedBy { it!!.type }
            return orderedPatients
        }
}